[book] Merge lookup arguments for normal and short variants (#162)
zcash/orchard: book/src/design/circuit/gadgets/decomposition.md

Strict mode constrains the running sum output $z_{W}$ to be zero, thus range-constraining the field element to be within $W \cdot K$ bits.

In strict mode, we are also assured that $z_{W-1} = k_{W-1}$ gives us the last window in the decomposition.
## Lookup decomposition
This gadget makes use of a $K$-bit lookup table to decompose a field element $\alpha$ into $K$-bit words. Each $K$-bit word $k_i = z_i - 2^K \cdot z_{i+1}$ is range-constrained by a lookup in the $K$-bit table.

The region layout for the lookup decomposition uses a single advice column $z$, and two selectors $q_{lookup}$ and $q_{running}.$
$$
\begin{array}{|c|c|c|}
\hline

>     z    & q_{lookup} & q_{running} \\\hline

Reviewer: ```suggestion
    z    & q_{lookup} & q_{running} \\\hline
\hline
```

Reviewer 2: ```suggestion
    z    & q_mathit{lookup} & q_mathit{running} \\\hline
```

  z_0    &     1      &       1     \\\hline
  z_1    &     1      &       1     \\\hline
\vdots   &   \vdots   &     \vdots  \\\hline
z_{n-1}  &     1      &       1     \\\hline
z_n      &     0      &       0     \\\hline
\end{array}
$$
### Short range check
Using two $K$-bit lookups, we can range-constrain a field element $\alpha$ to be $n$ bits, where $n \leq K.$ To do this:

1. Constrain $0 \leq \alpha < 2^K$ to be within $K$ bits using a $K$-bit lookup.
2. Constrain $0 \leq \alpha \cdot 2^{K - n} < 2^K$ to be within $K$ bits using a $K$-bit lookup.

The short variant of the lookup decomposition introduces a $q_{bitshift}$ selector. The same advice column $z$ has here been renamed to $\textsf{word}$ for clarity:
$$
\begin{array}{|c|c|c|c|}
\hline

> \textsf{word} & q_{lookup} & q_{running} & q_{bitshift} \\\hline

Reviewer: ```suggestion
\textsf{word} & q_{lookup} & q_{running} & q_{bitshift} \\\hline
\hline
```

Reviewer 2: ```suggestion
\textsf{word} & q_\mathit{lookup} & q_\mathit{running} & q_\mathit{bitshift} \\\hline
\hline
```

\alpha        &     1      &      0      &       0      \\\hline
\alpha'       &     1      &      0      &       1      \\\hline

> 2^{K-n}       &     0      &      0      &       0      \\\hline

Reviewer 2: If we used two advice columns here, then we could put 2<sup>K-n</sup> in the first row of the second column, and then we'd be able to refer to all three cells using only *cur* and *next* queries. I don't want to delay this PR getting in, so this can be a later optimization if and when we try to optimize other gates to only use *cur* and *next*.

\end{array}
$$

> where $\alpha' = \alpha \cdot 2^{K - n}.$ Note that $2^{K-n}$ is assigned to a fixed column at keygen, and copied in at proving time. This is used in the gate enabled by the $q_{bitshift}$ selector to check that $\alpha$ was shifted correctly:

Reviewer 2: ```suggestion
where $\alpha' = \alpha \cdot 2^{K - n}.$ Note that $2^{K-n}$ is assigned to a fixed column at keygen, and copied in at proving time. This is used in the gate enabled by the $q_\mathit{bitshift}$ selector to check that $\alpha$ was shifted correctly:
```

$$
\begin{array}{|c|l|}
\hline
\text{Degree} & \text{Constraint} \\\hline

>        2      & q_{bitshift} \cdot (\alpha' - (\alpha \cdot 2^{K - n})) \\\hline

Reviewer 2: ```suggestion
       2      & q_\mathit{bitshift} \cdot (\alpha' - (\alpha \cdot 2^{K - n})) \\\hline
```

\end{array}
$$

### Combined lookup expression

> Since the lookup decomposition and its short variant both make use of the same lookup table, we can combine their lookup input expressions into a single one:

Reviewer: ```suggestion
Since the lookup decomposition and its short variant both make use of the same lookup table, we combine their lookup input expressions into a single one:
```

> $$q_{lookup} \cdot \left(q_{running} \cdot (z_i - 2^K \cdot z_{i+1}) + (1 - q_{running}) \cdot \textsf{word} \right).$$

Reviewer: `word` is undefined. I haven't looked at the impl PR yet, but I presume that this should be:
```suggestion
$$q_{lookup} \cdot \left(q_{running} \cdot (z_i - 2^K \cdot z_{i+1}) + (1 - q_{running}) \cdot \textsf{word} \right)$$

where $z_i$ and $\textsf{word}$ are the same cell (but distinguished here for clarity of usage).
```

Reviewer 2: ```suggestion
$$q_\mathit{lookup} \cdot \left(q_\mathit{running} \cdot (z_i - 2^K \cdot z_{i+1}) + (1 - q_\mathit{running}) \cdot \textsf{word} \right)$$

where $z_i$ and $\textsf{word}$ are the same cell (but distinguished here for clarity of usage).
```

## Short range decomposition

> For a short range (for instance, $2^K \leq 8$), we can range-constrain each word using a degree-$2^K$ polynomial constraint instead of a lookup: $$\texttt{range\_check(word, range)} = \texttt{word} \cdot (1 - \texttt{word}) \cdots (\texttt{range} - 1 - \texttt{word}).$$

Reviewer 2: ```suggestion
For a short range (for instance, $[0, m)$ where $m \leq 8$), we can range-constrain each word using a degree-$m$ polynomial constraint instead of a lookup: $$\texttt{range\_check(word, range)} = \texttt{word} \cdot (1 - \texttt{word}) \cdots (\texttt{range} - 1 - \texttt{word}).$$
```

\hline
\textsf{word} & q_\mathit{lookup} & q_\mathit{running} & q_\mathit{bitshift} \\\hline
\hline
\alpha        &     1      &      0      &       0      \\\hline
\alpha'       &     1      &      0      &       1      \\\hline
2^{K-n}       &     0      &      0      &       0      \\\hline
\end{array}
$$

where $\alpha' = \alpha \cdot 2^{K - n}.$ Note that $2^{K-n}$ is assigned to a fixed column at keygen, and copied in at proving time. This is used in the gate enabled by the $q_\mathit{bitshift}$ selector to check that $\alpha$ was shifted correctly:
$$
\begin{array}{|c|l|}
\hline
\text{Degree} & \text{Constraint} \\\hline
       2      & q_\mathit{bitshift} \cdot (\alpha' - (\alpha \cdot 2^{K - n})) \\\hline
\end{array}
$$

### Combined lookup expression
Since the lookup decomposition and its short variant both make use of the same lookup table, we combine their lookup input expressions into a single one:

$$q_\mathit{lookup} \cdot \left(q_\mathit{running} \cdot (z_i - 2^K \cdot z_{i+1}) + (1 - q_\mathit{running}) \cdot \textsf{word} \right)$$

where $z_i$ and $\textsf{word}$ are the same cell (but distinguished here for clarity of usage).

## Short range decomposition

> For a short range (for instance, $[0, m)$ where $m \leq 8$), we can range-constrain each word using a degree-$m$ polynomial constraint instead of a lookup: $$\texttt{range\_check(word, range)} = \texttt{word} \cdot (1 - \texttt{word}) \cdots (\texttt{range} - 1 - \texttt{word}).$$

Reviewer 2: Oops, I introduced m without spotting that there was already `range`.
